### PR TITLE
MAG-141191-23618: Fixing issue with Invalid customer id

### PR DESCRIPTION
**Pull request by shrma-smit**
### Description (*)

Invalid customer ID issue coming on the checkout while placing the order. This issue comes randomly for the customers who are logged in on the website. This issue is related to the customer quote, the table quote_address stays empty for this quote.

### Fixed Issues (if relevant)

1. magento/magento2#23618: Invalid address id at check out

### Manual testing scenarios (*)

1. Add a product in the cart and proceed to checkout.
2. Click on Place order.
3. This issue is coming for the logged in customers randomly on the checkout.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Quote/Model/QuoteAddressValidator.php [hunk 1]
```diff
@@ -127,6 +127,6 @@ public function validate(AddressInterface $addressData)
      */
     public function validateForCart(CartInterface $cart, AddressInterface $address): void
     {
-        $this->doValidate($address, $cart->getCustomerIsGuest() ? null : $cart->getCustomer()->getId());
+        $this->doValidate($address, $cart->getCustomer()->getId() ? $cart->getCustomer()->getId() : null);
```
Comment (aleron75) [inline app/code/Magento/Quote/Model/QuoteAddressValidator.php]: Hello @sumitCueBlocks, thanks for your submission!
In this case, to reduce code redundancy, you can use the `?:` (Elvis) operator, thus the code would become:

`$this->doValidate($address, $cart->getCustomer()->getId() ?: null);`

My best

Comment (shrma-smit) [reply app/code/Magento/Quote/Model/QuoteAddressValidator.php]: Hi @aleron75,

Thank you for your suggestion, I've updated the function code as you've suggested to reduce code redundancy.

Regards
Sumit

**Review by aleron75 — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @aleron75, thank you for the review.
[ENGCOM-7215](https://jira.corp.magento.com/browse/ENGCOM-7215) has been created to process this Pull Request
:eight_spoked_asterisk: @aleron75, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by magento-engcom-team** [comment]
@shrma-smit thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by aleron75** [comment]
Hello @shrma-smit, 

due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html), all code must be covered by tests. 

For this specific case, you should cover your fix by automated tests with the scenario which leads to an issue. 
Tests should fail on the mainline and pass with your fix.

To answer the question _"which kind of tests should I write"_, the answer is:
pick the most lightweight (execution time-wise) test type that will provide sufficient coverage.

Since in Magento codebase there is already a unit test for this class, I would recommend adding a test case in `\Magento\Quote\Test\Unit\Model\Quote\Address\ValidatorTest`. 

Thank you again!

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
https://github.com/magento/magento2/pull/26637#issuecomment-608299232

**Review by shrma-smit — COMMENTED** [review]
Changes reviewed.

**Comment by VladimirZaets** [comment]
Hi @shrma-smit. Thanks for collaboration. As @aleron75 already mentioned, due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. Please cover your fix by automated tests.

**Comment by j-lloyd-slc** [comment]
This seems to be an ongoing problem in 2.3.4 because it assumes that if `quote`.`customer_is_guest` = 1, `quote`.`customer_id` = NULL.

I have found that this is not so, and when it is not, the subsequent validation fails and blocks checkout.

Granted, the root of the problem is that quote records are being populated with contradictory data. (Actually, the root of the problem is that quote records have two columns that can contradict each other at all). But as things are, this dysfunctional validation leaves customers in a state that they can not intuitively recover from.

**Comment by srenon** [comment]
@engcom-Alfa @leron75 using this PR will still cause data inconsistency 

For example there will not be a hyperlink to go directly to the customer record from sales order view
image

image

I think it would be better to set the correct values before submitting  a quote 

vendor/magento/module-quote/Model/QuoteManagement.php

Index: vendor/magento/module-quote/Model/QuoteManagement.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
---  vendor/magento/module-quote/Model/QuoteManagement.php	(date 1589668946000)
+++  vendor/magento/module-quote/Model/QuoteManagement.php	(date 1589668946000)
@@ -494,6 +494,12 @@
      */
     protected function submitQuote(QuoteEntity $quote, $orderData = [])
     {
+        if ($quote->getCustomer()->getId()) {
+            $quote->setCustomerIsGuest(0);
+        } else {
+            $quote->setCustomerIsGuest(1);
+        }
+
         $order = $this->orderFactory->create();
         $this->submitQuoteValidator->validateQuote($quote);
         if (!$quote->getCustomerIsGuest()) {

**Comment by engcom-Kilo** [comment]
@shrma-smit 
Can not reproduce this bug.

Please update descriptions,
add step by step instruction for reproduce.

**Comment by engcom-Kilo** [comment]
Hi @shrma-smit, I'm closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for your collaboration.

**Comment by jas8522** [comment]
I have reproduced this bug after upgrading from 2.2.9 to 2.4.0. It likely only occurs when there has been some kind of inconstency in the database, so how are you going to expect anyone to just reproduce it normally? Shouldn't the code provided here *still* be applied just in case anyone else experiences such an inconsistency?

**Comment by jackrevate** [comment]
Still happens in 2.4.3-p1 - unreal the amount of bugs with Magento. I just want a basic functioning Ecomm store! 

Doing the below fixed it, but is there anything else that needs to be done?

https://github.com/magento/magento2/pull/26637/files

File code to change:
app/code/Magento/Quote/Model/QuoteAddressValidator.php

Find:
$this->doValidate($address, $cart->getCustomerIsGuest() ? null : $cart->getCustomer()->getId());

Replace with:
$this->doValidate($address, $cart->getCustomer()->getId() ?: null);

**Comment by alitopaloglu** [comment]
This happens when your cart's

- customer_id is not null
- but customer_is_guest=1

so, this should not happen if your quote data is consistent. 

Better to search for any other bug in 3rd parties that causes this inconsistency.

**Pull request closed without merging**